README.md
# Homebridge-plugin-samsung-air-conditioner

[![npm version](https://badge.fury.io/js/homebridge-plugin-samsung-air-conditioner.svg)](https://badge.fury.io/js/homebridge-plugin-samsung-air-conditioner)

Homebridge plugin for controlling Samsung Air Conditioner working on port 2878. Allows to control AC with HomeKit and Siri.
If you have Samsung AC which operates on port 8888, check this plugin instead: https://github.com/cicciovo/homebridge-samsung-airconditioner

## Installation
1. Install [Homebridge](https://github.com/nfarina/homebridge).
2. Install this plugin by running `npm install -g homebridge-plugin-samsung-air-conditioner`.
3. Assign static IP address to your AC (check your router settings to do that).
4. Run `homebridge-samsung-ac-get-token <your ac's ip address>` in terminal and follow instructions.
5. Update your Homebridge `config.json`. Check `config-sample.json` for reference. 
    - Required parameters:
        - `accessory` - always "Samsung Air Conditioner"
        - `name` - Name of your device
        - `ip_address` - IP address of air conditioner
        - `mac` - MAC address of air conditioner in format `AA:BB:CC:DD:EE:FF` or `AA-BB-CC-DD-EE-FF`
        - `token` - token returned by `homebridge-samsung-ac-get-token <your ac's ip address>`
    - Optional parameters:
        - `log_socket_activity` - `true`/`false` (default `false`). If `true` then logs additional raw data to console
        - `keep_alive` - dictionary with keep alive settings:
            - `enabled` - `true`/`false` (default `true`). If `true` then enables keep alive on underlying socket
            - `initial_delay` - milliseconds as integer (default `10000`). Time which needs to pass after last write to socket before sending first keep alive packet
            - `interval` - milliseconds as integer (default `10000`). Time between keep alive packes
            - `probes` - integer (default `10`). Number of keep alive packets to fails before treating connection as closed

## Features
- Turning AC on and off
- Getting and setting target temperature
- Getting current temperature
- Getting and setting mode
- Getting and setting swing mode
- Getting and setting wind level
- Reacting to changes made by using AC's remote

## Confirmed compatibility list (model numbers)
- AR12HSSFAWKNEU
- AR18HSFSAWKNEU
- AR24FSSSBWKN
- AR12FSSEDWUNEU

If your device's number is not on the list but you have tested it and it works, please make a PR with your device's number.

## Acknowledgment
This project is heavily based on awesome work of CloCkWeRX - https://github.com/CloCkWeRX/node-samsung-airconditioner


CHANGELOG.md
# Changelog

## 2.0.0 - 18.11.2018
### Added
- Support for controlling oscillation
### Improved
- Improve response time by introducing local cache for AC state
- Improve error handling
- Improve reconnecting after power loss, networking issues etc.
- Improve project structure

## 1.0.3 - 25.10.2018
### Fixed
- Fix for not reconnecting after socket closes

## 1.0.2 - 14.10.2018
### Fixed
- Fix for connection error caused by too weak DH key

## 1.0.1 - 09.10.2018
### Fixed
- Fix wrong command name in documentation
- Fix crash during reconnection

## 1.0.0 - 08.10.2018
Initial release


lib/air-conditioner-api.js
const events = require('events');
const util = require('util');
const tls = require('tls');
const carrier = require('carrier');
const fs = require('fs');
const path = require('path');
const shortid = require('shortid');
const keepAlive = require('net-keepalive');

const port = 2878;

function AirConditionerApi(ipAddress, duid, token, log, logSocketActivity, keepAliveConfig) {
    this.ipAddress = ipAddress;
    this.duid = duid;
    this.token = token;
    this.log = log;
    this.logSocketActivity = logSocketActivity;

    const defaultKeepAliveConfig = {
        "enabled": true,
        "initial_delay": 10000,
        "interval": 10000,
        "probes": 10
    }

    this.keepAliveConfig = Object.assign({}, defaultKeepAliveConfig, keepAliveConfig);

    log('Keep alive config:', this.keepAliveConfig);

    this.authenticated = false;
};

AirConditionerApi.prototype = {
    connect: function () {
        this.log('Connecting...');

        this.controlCallbacks = {};

        const pfxPath = path.join(__dirname, '../res/ac14k_m.pfx')

        const options = {
            pfx: fs.readFileSync(pfxPath),
            port: port,
            host: this.ipAddress,
            rejectUnauthorized: false,
            ciphers: 'HIGH:!DH:!aNULL'
        };

        this.socket = tls.connect(options, function () {
            this.log('Connected');

            this.socket.setKeepAlive(this.keepAliveConfig.enabled, this.keepAliveConfig.initial_delay);
            keepAlive.setKeepAliveInterval(this.socket, this.keepAliveConfig.interval);
            keepAlive.setKeepAliveProbes(this.socket, this.keepAliveConfig.probes);

            // All responses from AC are received here as lines
            carrier.carry(this.socket, this._readLine.bind(this));
        }.bind(this));

        this.socket
            .on('end', this._connectionEnded.bind(this))
            .on('close', this._connectionClosed.bind(this))
            .on('error', this._errorOccured.bind(this));
    },

    deviceControl: function (key, value, callback) {
        if (!this.authenticated) {
            callback(new Error('Connection not established'));

            return;
        }

        // Create id for callback. It will be passed to request and returned by AC in response
        // It allows us to match callbacks to responses received in `carrier.carry` callback above
        const id = shortid.generate()

        if (!!callback) {
            this.controlCallbacks[id] = callback;
        }

        this._send(
            '<Request Type="DeviceControl"><Control CommandID="' + id + '" DUID="' + this.duid + '"><Attr ID="' + key + '" Value="' + value + '" /></Control></Request>'
        );
    },

    _send: function (line) {
        if (this.logSocketActivity) { this.log('Write:', line); }

        this.socket.write(line + "\r\n");
    },

    _readLine: function (line) {
        if (this.logSocketActivity) { this.log('Read:', line); }
        
        if (line.match(/Update Type="InvalidateAccount"/)) { // Returned in the beginning of connection. We need to send auth request with token.
            this._handleInvalidateAccount();
        } else if (line.match(/Response Type="AuthToken" Status="Okay"/)) { // Auth success
            this._handleAuthSuccessResponse();
        } else if (line.match(/Update Type="Status"/)) { // Status update received - AC sends them when some setting is changed via remote. 
            this._handleDeviceStatusUpdate(line);
        } else if (line.match(/Response Type="DeviceState" Status="Okay"/)) { // Status response received - AC sends it after receiving request for status
            this._handleDeviceStateResponse(line);
        } else if (line.match(/Response Type="DeviceControl" Status="Okay"/)) { // Control confirmation received - AC sends it to confirm control success
            this._handleDeviceControlResponse(line);
        };
    },

    _handleInvalidateAccount: function() {
        this.log("Auth request received - Authenticating...");

        this._send('<Request Type="AuthToken"><User Token="' + this.token + '"/></Request>');
    },

    _handleAuthSuccessResponse: function() {
        this.log("Authentication succeeded");

        this.authenticated = true;

        this.log("Requesting full state summary...");

        // Request full state summary;
        this._send('<Request Type="DeviceState" DUID="' + this.duid + '"></Request>');
    },

    _handleDeviceStateResponse: function (line) {
        this.log("Full state summary received");

        const attributes = line.split("><");
        const state = {};
        attributes.forEach(function (attr) {
            if ((matches = attr.match(/Attr ID="(.*)" Type=".*" Value="(.*)"/))) {
                const id = matches[1];
                state[id] = matches[2];
            }
        }.bind(this));

        this.emit('stateUpdate', state);
    },

    _handleDeviceControlResponse: function (line) {
        if ((matches = line.match(/CommandID="(.*)"/))) {
            id = matches[1];

            if (!this.controlCallbacks[id]) return;
            callback = this.controlCallbacks[id];
            delete (this.controlCallbacks[id]);

            callback(null);
        }
    },

    _handleDeviceStatusUpdate: function(line) {
        if ((matches = line.match(/Attr ID="(.*)" Value="(.*)"/))) {
            const state = {};
            state[matches[1]] = matches[2];

            this.emit('stateUpdate', state);
        }
    },

    _errorOccured: function(error) {
        this.log('Error occured:', error.message);

        // Error all callbacks
        Object.keys(this.controlCallbacks).forEach(function(id) {
            this.controlCallbacks[id](error);
        }.bind(this));

        this.controlCallbacks = {};
    },

    _connectionEnded: function () {
        this.log('Connection ended');
    },

    _connectionClosed: function (hadError) {
        this.authenticated = false;

        this.log('Connection closed' + (hadError ? ' because error occured' : ''));
        this.log('Trying to reconnect in 5s...');

        setTimeout(this.connect.bind(this), 5000);
    }
};

util.inherits(AirConditionerApi, events.EventEmitter);

module.exports = AirConditionerApi;